### DISCUSSION TITLE
Computing descriptor in orb_impl

### DISCUSSION THREAD

**Discussion by Ashmit101** [stella-cv/stella_vslam#556] [Q&A] [answered]
Hey, 

I was trying to use the cuda functions to extract the ORB features from the frames. As I was browsing the code, I discovered that simple "detectAndCompute" functions were not used. In orb_impl, there was a complex way of calculating the descriptor of a single keypoint.

Why was it done so? How will the descriptors and keypoints be different if we use `detectAndCompute` function of OpenCV?

**Comment by jjd9** [comment, marked as answer]
> How will the descriptors and keypoints be different if we use detectAndCompute function of OpenCV?

OpenCV and stella_vslam actually use the same algorithm (i think its even the same code implementation from Willow Garage) for computing the feature descriptor.
The difference is in the feature keypoint extraction. stella_vslam uses a sort of adaptive non-maximum suppression to select a set of keypoints that are distinctive in the image but also spatially distributed (or spread out). Whereas opencv just returns all the keypoints that are detected in order of their feature response in the image.
So if you ran the opencv ORB function and the stella_vslam orb function on an image and compared the output, you would likely see that the stella_vslam function returns points that are more spread out over the image. And if you compared feature descriptors at overlapping or nearby keypoints, they should be very similar if not identical.

> Why was it done so?

idk the original reason for this decision, but my intuition is that since OpenCV is a general purpose library, it just gives you the keypoints and lets you decide what to do with them. Whereas stella_vslam is meant for SLAM where processing time is at a premium. Since SLAM is meant to run in real-time, there is a limit on the number of features that can be processed and matched in each frame. So it is helpful to have as well-chosen features and as fine-grained control over their quantity as possible. Hence the clever way that stella_vslam adaptively selects strong, well-distributed FAST features.
